src/CustomFunction.ts

```diff
@@ -0,0 +1,211 @@
+      token,
+      outputs: params.outputs || {},
+      function_execution_id: functionExecutionId,
+    });
+  }
+
+  /**
+ * Factory for `fail()` utility
+ * @param args function_executed event
+ */
+  public static createFunctionFail(context: Context, client: WebClient): FunctionFailFn {
+    const token = selectToken(context);
+
+    return (params: Parameters<FunctionFailFn>[0]) => {
+      const { error } = params ?? {};
+      const { functionExecutionId } = context;
+
+      return client.functions.completeError({
+        token,
+        error,
+        function_execution_id: functionExecutionId,
+      });
+    };
+  }
+}
+
+/** Helper Functions */
+export function validate(callbackId: string, middleware: CustomFunctionExecuteMiddleware): void {
+  // Ensure callbackId is valid
+  if (typeof callbackId !== 'string') {
+    const errorMsg = 'CustomFunction expects a callback_id as the first argument';
+    throw new CustomFunctionInitializationError(errorMsg);
+  }
+
+  // Ensure middleware argument is either a function or an array
+  if (typeof middleware !== 'function' && !Array.isArray(middleware)) {
+    const errorMsg = 'CustomFunction expects a function or array of functions as the second argument';
+    throw new CustomFunctionInitializationError(errorMsg);
+  }
+
+  // Ensure array includes only functions
+  if (Array.isArray(middleware)) {
+    middleware.forEach((fn) => {
+      if (!(fn instanceof Function)) {
+        const errorMsg = 'All CustomFunction middleware must be functions';
+        throw new CustomFunctionInitializationError(errorMsg);
+      }
+    });
+  }
+}
+
+/**
+ * `processFunctionMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processFunctionMiddleware(
+  args: AllCustomFunctionMiddlewareArgs,
+  middleware: CustomFunctionMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(
+      callbacks, args, context, client, logger,
+      async () => lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isFunctionEvent(args: AnyMiddlewareArgs): args is AllCustomFunctionMiddlewareArgs {
+  return VALID_PAYLOAD_TYPES.has(args.payload.type);
+}
+
+function selectToken(context: Context): string | undefined {
+  // If attachFunctionToken = false, fallback to botToken or userToken
+  return context.functionBotAccessToken ? context.functionBotAccessToken : context.botToken || context.userToken;
+}
+
+/**
+ * `enrichFunctionArgs()` takes in a function's args and:
+ *  1. removes the next() passed in from App-level middleware processing
+ *    - events will *not* continue down global middleware chain to subsequent listeners
```

Comment:
🤔 Is removing this needed for correct function handling? I can imagine adding middleware after the function listener handler to do something specific to functions after some response like so:

```js
app.function("example_callback", filterInputs, async ({ inputs, complete, next }) => {
  await complete({ outputs: inputs });
  await next();
}, saveResponseToDatabase);
```

It's not super clear that this is possible in docs - looking into this now - but I'm finding that listener middleware can go in any order, unrelated to the main listener handler.

Things do get interesting if `functionBotAccessToken: true` is used in that example though since the attached `xwfp` `token` is revoked soon after `complete`, meaning `saveResponseToDatabase` will get a `token_revoked` error if it tries using it...

But I think that error is better than swapping for the original `bot` or `user` token too since those tokens can be found again. I'm curious about the middleware handling most though! Leaning towards having `processMiddleware` handle callbacks and middleware without change if that's possible :eyes:

Reply:
Global middleware should also run before these listeners too, but `app.error` might catch thrown errors if that happens. Thinking this is all fine as is!

Reply:
This behavior works as expected, and is modeled off of the original `WorkflowStep` class. `next` is still available for developers to use within the callback in the case of additional middleware that they've included, but it's removed here so as not to allow the event to fall further down.

Admittedly, this is overkill, as `WorkflowStep` featured events that could be handled by more than one handler (ie, `view_submission`). We don't have a downside of keeping it, and it makes it extensible if and when there _is_ overlap, but if folks feel strongly about it, we can remove it (though it will need to be re-tested).